### PR TITLE
Basic Improvements to Sandman Shell

### PR DESCRIPTION
**Pull request by serene-meadow**
## Sandman Shell | Basic Improvements

Previously, the Sandman shell did not echo the user keyboard input back to the terminal.
Also, entering an empty string would crash the program. This pull request fixes those issues.

The shell was changed to display a designated region for echoing user keyboard input.

The `NCurses` namespace was added to incapsulate functionality and state relating to NCurses.

### `LoggerAddMessage` | [GCC Format Attribute](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-format-function-attribute)

The `LoggerAddMessage` functions in *logger.h* are in the style of `std::printf`.

By adding the GNU attribute *format* to these functions, this allows the GCC compiler to make sure
the arguments match the correct format specifiers.

```C++
[[gnu::format(printf, 1, 2)]] bool LoggerAddMessage(char const* p_Format, ...);

[[gnu::format(printf, 1, 0)]] bool LoggerAddMessage(char const* p_Format, std::va_list& p_Arguments);
```

The first argument of the `gnu::format` attribute is the style of format function to use.
`LoggerAddMessage` is like `printf`, so that is the first argument.

The second argument of the `gnu::format` attribute is the index (starting from one) of the format string.
The first argument of `LoggerAddMessage` is the format string, so this is `1`.

The third argument of the `gnu::format` attribute is the index (starting from one) of the ellipsis denoting the arguments
to be interpolated into the format string, so the third argument is `2` for the former overload of `LoggerAddMessage`. However, if there is no ellipsis, like in the case of the `std::va_list` overload of `LoggerAddMessage`, the third argument should be `0`.

> A side effect of annotating `LoggerAddMessage` with `gnu::format` is that the expression
`LoggerAddMessage("")` triggers a warning from GCC as being erroneous. So, `LoggerAddEmptyLine()`
was added which accomplishes the same thing as `LoggerAddMessage("")`.

Like `LoggerAddMessage`, `Input::CloseDevice` in *input.h* was also annotated with `gnu::format`.
```C++
[[gnu::format(printf, 1+2, 1+3)]] void CloseDevice(bool p_WasFailure, char const* p_Format, ...);
```

Note that, for `Input::CloseDevice`, the second and third arguments of `gnu::format` are offset by one because
`Input::CloseDevice` takes an implicit first argument `this`.

### `LoggerPrintArgs(ParamsT const&... p_Arguments)`

```C++
template <typename... ParamsT>
inline bool LoggerPrintArgs(ParamsT const&... p_Arguments)
{
	return LoggerAddMessage("%s", (std::ostringstream() << ... << p_Arguments).str().c_str());
}
```

This function uses an output string stream `std::ostringstream` to convert its arguments to a string which is passed to `LoggerAddMessage`. So, this function can be used as a quick print statement when debugging, or as shorthand for when you would like to make use of `std::ostringstream` to format a string and log it.

Note, it does construct and destroy a `std::ostringstream` every call.

## Changes | *sandman/source/*

All the changes are in *sandman/source/*.

### *CMakeLists.txt*
* Added `ncurses_context.cpp` to the list of C++ source files `SOURCE_FILES`.
	* &check; Maintained alphabetical order.

### *Makefile.am*
* Added `ncurses_context.cpp` to the list of C++ source files `sandman_SOURCES`.
	* &check; Maintained alphabetical order.

### *command.cpp*
* Now using `std::tolower(CharT, const locale&)` from `<locale>` instead of `std::tolower(int)` from `<cctype>`.
	* `std::tolower(int)` is undefined for signed characters, while `std::tolower(CharT, const locale&)` doesn't have this restriction.
		* Note: the signedness of `char` depends on the compiler and the target platform, so the safe way to use `std::tolower(int)` is to cast the character to `unsigned char` first.
	* `std::tolower(int)` depends on the currently installed C locale, while the locale for `std::tolower(CharT const locale&)` can be specified for more consistent behavior.
* Fix bug: sandman shell crashes when receiving empty string.
	* Reason: the empty string `""` gets passed to `std::stoul` which then throws an exception.
	* Fix: use `std::from_chars` to attempt to parse the string into a number. Only set `l_Token.m_Type = CommandToken::TYPE_INTEGER` on success. This handles the empty string case; it will be a failed parse instead of throwing an exception.

### *command.h*
* `CommandToken::m_Parameter` is now `unsigned int` instead of `int`. This is to match with `unsigned int p_DurationPercent` in the function `Control::SetDesiredAction`.

### *control.cpp*
* Bug fix: incorrect format string when calling `LoggerAddMessage` in `Control::Process`.
	* `m_State` is `enum Control::State`, but was incorrectly formatted with `%s`, now changed to `%d`.
	* `m_Name` is `char[32]`, but was incorrectly formatted with `%d`, now changed to `%s`.
	* This bug was identified using the `gnu::format` attribute on `LoggerAddMessage`.
* Replaced `LoggerAddMessage("")` with `LoggerAddEmptyLine()`.

### *input.cpp*
* Replaced `LoggerAddMessage("")` with `LoggerAddEmptyLine()`.

### *input.h*
* Annotated `Input::CloseDevice` with `gnu::format`.
	* ```C++
		[[gnu::format(printf, 1+2, 1+3)]] void CloseDevice(bool p_WasFailure, char const* p_Format, ...);
		```

### *logger.cpp*
* Include `"ncurses_context.h"` instead of `<ncurses.h>`.
* Print to the logging window defined in `"ncurses.cpp"` instead of the NCurses standard screen window `stdscr`.
	* Before:
		```C++
		// These print to NCurses's `stdscr`.
		addstr(l_LogStringBuffer);
		addch('\n');
		refresh();
		```
	* After:
		```C++
		WINDOW* const l_Window = NCurses::GetLoggingWindow();
		waddstr(l_Window, l_LogStringBuffer);
		waddch(l_Window, '\n');
		wrefresh(l_Window);
		```

### *logger.h*
* Annotate `printf`-like functions with `gnu::format` attribute.
	* ```C++
		[[gnu::format(printf, 1, 2)]] bool LoggerAddMessage(char const* p_Format, ...);
		```
	* ```C++
		[[gnu::format(printf, 1, 0)]] bool LoggerAddMessage(char const* p_Format, std::va_list& p_Arguments);
		```

* **NEW FUNCTION** definition `LoggerAddEmptyLine`
	* `LoggerAddEmptyLine()` is semantically equivalent to `LoggerAddMessage("")`.

* **NEW FUNCTION** definition `LoggerPrintArgs`
	* &check; Type safe.
	* Logs variable number of arguments.
		```C++
		LoggerPrintArgs("Integer: ", 10, ", Boolean: ", std::boolalpha, true, ", Floating Point: ", 5.0);
		```
	* No format string.
	* Useful for quick debugging, but constructs and destroys a `std::ostringstream` every call.

### *main.cpp*
* Replaced the initialization of NCurses with the function call `NCurses::Initialize()` in the `Initialize` function.
* Replaced `endwin()` with `NCurses::Uninitialize()` in the `Unitialize` function.
* In the function `ProcessKeyboardInput`, echo the user input to the input window.
* Reformatted this file with ClangFormat; this changed much of the whitespace. 😵‍💫

### *mqtt.cpp*
* Replaced `LoggerAddMessage("")` with `LoggerAddEmptyLine()`;

### **NEW FILE** *ncurses_context.cpp*
* `static WINDOW* s_LoggingWindow`
	* This window is where messages from the logger are written to.
* `static WINDOW* s_InputWindow`
	* This window is where user input is echoed to.
* `s_LoggingWindow` and `s_InputWindow` pointers are private to the `NCurses` namespace so that only the `NCurses` namespace can set them.

### **NEW FILE** *ncurses_context.h*
* Defines the new `NCurses` namespace.

### *schedule.cpp*
* Replaced `LoggerAddMessage("")` with `LoggerAddEmptyLine()`.

### CODE REVIEW

File: sandman/source/logger.h [hunk 1]
```diff
@@ -42,5 +43,42 @@ bool LoggerAddMessage(char const* p_Format, ...);
...
 
+/**
+ * @brief Log an empty line. (The logger still prints the date and time.)
+ * 
+ * @note `LoggerAddMessage("")` triggers a warning `zero-length gnu_printf format string
+ * [-Werror=format-zero-length]`; this function accomplishes the same without triggering a warning.
+ * 
+ * @return `true` if successful, `false` otherwise
+ */
+inline bool LoggerAddEmptyLine()
```
Comment (shawn-lindberg) [inline sandman/source/logger.h:57]: I'm not in love with pragmas to solve this. I think I would rather implement a real function for this.

File: sandman/source/main.cpp [hunk 2]
```diff
@@ -287,18 +273,66 @@ static void Uninitialize()
 //
 // returns:		True if the quit command was processed, false otherwise.
 //
-static bool ProcessKeyboardInput(char* p_KeyboardInputBuffer, unsigned int& p_KeyboardInputBufferSize, 
-	unsigned int const p_KeyboardInputBufferCapacity)
+static bool ProcessKeyboardInput(char* p_KeyboardInputBuffer,
```
Comment (shawn-lindberg) [inline sandman/source/main.cpp:276]: I'd like to move most of or maybe this entire function into NCurses.

File: sandman/source/ncurses_context.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			nodelay(p_Window, TRUE);
+
+			// Don't set a timer to wait for the next character when interpreting an escape sequence.
+			notimeout(p_Window, TRUE);
+		}
+	}
+
+} // namespace NCurses
+
+void NCurses::Initialize()
```
Comment (shawn-lindberg) [inline sandman/source/ncurses_context.cpp:66]: These functions should also be inside of the namespace scope.

File: sandman/source/ncurses_context.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	WINDOW* GetInputWindow()
+	{
+		return s_InputWindow;
+	}
+
+	/*
+		Configure a window with "sensible" defaults.
+	*/
+	static void DefaultConfigureWindow(WINDOW* const p_Window)
```
Comment (shawn-lindberg) [inline sandman/source/ncurses_context.cpp:25]: Naming it this way makes me think that this is the default function, but I suspect it's the only one, so it's clear if it's named ConfigureWindowDefaults

File: sandman/source/ncurses_context.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		nonl();
+	}
+
+	// Configure standard screen window.
+	DefaultConfigureWindow(stdscr);
+
+	// The input window will have a height of 3.
+	static constexpr int s_InputWindowRowCount{ 3 };
+
+	/* configure logging window */
```
Comment (shawn-lindberg) [inline sandman/source/ncurses_context.cpp:103]: Prefer line comments especially for single line comments.

**Review by shawn-lindberg — CHANGES_REQUESTED** [review]
Overall looks really good. I would like to rename ncurses_context

File: sandman/source/ncurses_context.cpp (hunk 5 above)
Comment (serene-meadow) [reply sandman/source/ncurses_context.cpp]: Fixed. Now there isn't a single `/*` in `source`.

**Comment by serene-meadow** [comment]
You can [view files changed with tab size 3.](https://github.com/shawn-lindberg/sandman/pull/45/files?ts=3)

File: sandman/source/command.cpp [hunk 6]
```diff
@@ -427,21 +428,21 @@ void CommandTokenizeString(std::vector<CommandToken>& p_CommandTokens,
...
-				l_IsNumeric = l_IsNumeric && std::isdigit(l_Character);
-			}
-			
-			if (l_IsNumeric == true)
+			// Attempt to parse the string into a number; save result into `l_Token.m_Parameter`.
+			auto const [endPointer, errorCode] = std::from_chars(l_TokenStringView.begin(),
+																				  l_TokenStringView.end(),
+																				  l_Token.m_Parameter);
+
+			if (/* successfully parsed to number */ errorCode == std::errc{0} and
```
[the remark this answers is not in the document]
Comment (serene-meadow) [reply sandman/source/command.cpp]: Renamed `ncurses_context` to `ncurses_ui`. I also considered `ncurses_user_interface`.

**Review by shawn-lindberg — APPROVED** [review]

**Pull request merged**